protowhat/sct_syntax.py

```diff
@@ -185,3 +201,108 @@ def create_sct_context(State, sct_dict, root_state=None):
+
+    Args:
+        parent_state: state of the host technology to derive the embedded state from
+        xstate: the State class of the embedded technology
+        state_args: extra arguments to pass to the constructor of the embedded state
+        highlight_offset: position of the embedded code in the student code
+
+    Returns:
+        an instance of xstate
+    """
+    # find all arguments the xstate constructor can handle
+    embedded_state_params = [
+        param
+        for cls in [xstate, *xstate.__bases__]
+        for param in inspect.signature(cls).parameters
+    ]
+
+    # gather the kwargs the xstate will be created with
+    args = {}
+
+    # copy all allowed arguments from the parent state
+    for arg in embedded_state_params:
+        if hasattr(parent_state, arg):
+            args[arg] = getattr(parent_state, arg)
+
+    # manually add / override arguments and
+    # configure the reporter to collaborate with the parent state reporter
+    args.update(
+        {
+            **(state_args or {}),
+            "reporter": Reporter(
+                parent_state.reporter, highlight_offset=highlight_offset or {}
+            ),
+        }
+    )
+
+    return xstate(**args)
+
+
+def create_embed_context(context: Chain, technology: str, **kwargs):
+    """
+    Create the globals that will be available when running the checks for the embedded technology.
+
+    Extra keyword arguments are passed to the constructor of the State for the embedded technology.
+
+    Args:
+        context: the Chain of the host technology
+            the checks for the embedded technology will use as starting point
+        technology: the name of the embedded technology (the x in xwhat)
+
+    Returns:
+        dict: the globals available to the SCT code for the embedded technology
+    """
+    parent_state = context._state
+
+    xwhat = __import__("{}what".format(technology))
+
+    xstate = xwhat.State.State
+
+    embedded_state = create_embed_state(parent_state, xstate, **kwargs)
+
+    return create_sct_context(
+        xstate, xwhat.sct_syntax.sct_dict, root_state=embedded_state
+    )
+
+
+def get_embed_chain_constructors(*args, **kwargs):
+    """
+    Get the chain constructors for the embedded technology.
+
+    This is a wrapper around create_embed_context
+    """
+    new_context = create_embed_context(*args, **kwargs)
+    return new_context["Ex"], new_context["F"]
+
+
+@contextmanager
+def embed_xwhat(*args, **kwargs):
+    """
+    This context manager temporarily updates the globals to be those for the embedded technology.
+    The context manager also returns the chain constructors for the embedded technology
+
+    This is a wrapper around create_embed_context
+    """
+    globals_backup = globals().copy()
+
+    new_context = create_embed_context(*args, **kwargs)
+    EmbeddedEx = new_context["Ex"]
+    EmbeddedF = new_context["F"]
+
+    globals().update(new_context)
+    yield EmbeddedEx, EmbeddedF
```

Comment:
Do we need to yield these? Since we're relying on updating the globals anyway? Might be the source of confusion at some point.